I will look at this tomorrow:
```
--- FAIL: TestObjectAPIPutObjectSingle (2.69s)
##[error]    object-api-putobject_test.go:177: Test 29: XL: Expected to pass, but failed with: error io: read/write on closed pipe.
##[error]    object-api-putobject_test.go:184: Test 29: XL: Expected to fail with error "<nil>", but instead failed with error "io: read/write on closed pipe" instead.
```

--- a/cmd/object-api-putobject_test.go
+++ b/cmd/object-api-putobject_test.go
@@ -172,22 +172,27 @@ func testObjectAPIPutObject(obj ObjectLayer, instanceType string, t TestErrHandl
 	}
 
 	for i, testCase := range testCases {
-		objInfo, actualErr := obj.PutObject(context.Background(), testCase.bucketName, testCase.objName, mustGetPutObjReader(t, bytes.NewReader(testCase.inputData), testCase.intputDataSize, testCase.inputMeta["etag"], testCase.inputSHA256), ObjectOptions{UserDefined: testCase.inputMeta})
+		in := mustGetPutObjReader(t, bytes.NewReader(testCase.inputData), testCase.intputDataSize, testCase.inputMeta["etag"], testCase.inputSHA256)
+		objInfo, actualErr := obj.PutObject(context.Background(), testCase.bucketName, testCase.objName, in, ObjectOptions{UserDefined: testCase.inputMeta})
 		if actualErr != nil && testCase.expectedError == nil {
 			t.Errorf("Test %d: %s: Expected to pass, but failed with: error %s.", i+1, instanceType, actualErr.Error())
+			continue
 		}
 		if actualErr == nil && testCase.expectedError != nil {
 			t.Errorf("Test %d: %s: Expected to fail with error \"%s\", but passed instead.", i+1, instanceType, testCase.expectedError.Error())
+			continue
 		}
 		// Failed as expected, but does it fail for the expected reason.
 		if actualErr != nil && actualErr != testCase.expectedError {
 			t.Errorf("Test %d: %s: Expected to fail with error \"%v\", but instead failed with error \"%v\" instead.", i+1, instanceType, testCase.expectedError, actualErr)
+			continue
 		}
 		// Test passes as expected, but the output values are verified for correctness here.
 		if actualErr == nil {
 			// Asserting whether the md5 output is correct.
 			if expectedMD5, ok := testCase.inputMeta["etag"]; ok && expectedMD5 != objInfo.ETag {
 				t.Errorf("Test %d: %s: Calculated Md5 different from the actual one %s.", i+1, instanceType, objInfo.ETag)
+				continue
 			}
 		}
 	}

--- a/cmd/test-utils_test.go
+++ b/cmd/test-utils_test.go
@@ -222,6 +222,8 @@ func initFSObjects(disk string, t *testing.T) (obj ObjectLayer) {
 // This makes it easy to run the TestServer from any of the tests.
 // Using this interface, functionalities to be used in tests can be made generalized, and can be integrated in benchmarks/unit tests/go check suite tests.
 type TestErrHandler interface {
+	Log(args ...interface{})
+	Logf(format string, args ...interface{})
 	Error(args ...interface{})
 	Errorf(format string, args ...interface{})
 	Failed() bool

--- a/pkg/hash/errors.go
+++ b/pkg/hash/errors.go
@@ -16,6 +16,8 @@
 
 package hash
 
+import "fmt"
+
 // SHA256Mismatch - when content sha256 does not match with what was sent from client.
 type SHA256Mismatch struct {
 	ExpectedSHA256   string
@@ -34,4 +36,13 @@ type BadDigest struct {
 
 func (e BadDigest) Error() string {
 	return "Bad digest: Expected " + e.ExpectedMD5 + " does not match calculated " + e.CalculatedMD5
+}
+
+type ErrSizeMismatch struct {
+	Want int64
+	Got  int64
+}
+
+func (e ErrSizeMismatch) Error() string {
+	return fmt.Sprintf("Size mismatch: got %d, want %d", e.Got, e.Want)
 }

--- a/pkg/hash/reader.go
+++ b/pkg/hash/reader.go
@@ -28,8 +28,6 @@ import (
 	sha256 "github.com/minio/sha256-simd"
 )
 
-var errNestedReader = errors.New("Nesting of Reader detected, not allowed")
-
 // Reader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
 // Reader verifies that the content of the io.Reader matches the expected checksums.
 type Reader struct {
@@ -51,7 +49,7 @@ func NewReader(src io.Reader, size int64, md5Hex, sha256Hex string, actualSize i
 	}
 
 	// Create empty reader and merge into that.
-	r := Reader{src: src}
+	r := Reader{src: src, size: -1, actualSize: -1}
 	return r.merge(size, md5Hex, sha256Hex, actualSize, strictCompat)
 }
 
@@ -69,7 +67,7 @@ func (r *Reader) Read(p []byte) (n int, err error) {
 
 	// At io.EOF verify if the checksums are right.
 	if err == io.EOF {
-		if cerr := r.Verify(); cerr != nil {
+		if cerr := r.verify(); cerr != nil {
 			return 0, cerr
 		}
 	}
@@ -122,9 +120,9 @@ func (r *Reader) SHA256HexString() string {
 	return hex.EncodeToString(r.sha256sum)
 }
 
-// Verify verifies if the computed MD5 sum and SHA256 sum are
+// verify verifies if the computed MD5 sum and SHA256 sum are
 // equal to the ones specified when creating the Reader.
-func (r *Reader) Verify() error {
+func (r *Reader) verify() error {
 	if r.sha256Hash != nil && len(r.sha256sum) > 0 {
 		if sum := r.sha256Hash.Sum(nil); !bytes.Equal(r.sha256sum, sum) {
 			return SHA256Mismatch{hex.EncodeToString(r.sha256sum), hex.EncodeToString(sum)}
@@ -141,13 +139,20 @@ func (r *Reader) Verify() error {
 // merge another hash into this one.
 // There cannot be conflicting information given.
 func (r *Reader) merge(size int64, md5Hex, sha256Hex string, actualSize int64, strictCompat bool) (*Reader, error) {
-	// Merge sizes.
-	if r.size <= 0 && size >= 0 {
-		r.size = size
-		if size > 0 {
-			r.src = io.LimitReader(r.src, size)
-		}
+	if r.bytesRead > 0 {
+		return nil, errors.New("internal error: Already read from hash reader")
 	}
+	// Merge sizes.
+	// If not set before, just add it.
+	if r.size < 0 && size >= 0 {
+		r.src = io.LimitReader(r.src, size)
+		r.size = size
+	}
+	// If set before and set now they must match.
+	if r.size >= 0 && size >= 0 && r.size != size {
+		return nil, ErrSizeMismatch{Want: r.size, Got: size}
+	}
+
 	if r.actualSize <= 0 && actualSize >= 0 {
 		r.actualSize = actualSize
 	}

--- a/pkg/hash/reader.go
+++ b/pkg/hash/reader.go
@@ -163,7 +163,7 @@ func (r *Reader) merge(size int64, md5Hex, sha256Hex string, actualSize int64, s
 		return nil, SHA256Mismatch{}
 	}
 
-	// If both are set, they must expect the same.
+	// If both are set, they must be the same.
 	if r.sha256Hash != nil && len(sha256sum) > 0 {
 		if !bytes.Equal(r.sha256sum, sha256sum) {
 			return nil, SHA256Mismatch{}

--- a/pkg/hash/reader_test.go
+++ b/pkg/hash/reader_test.go
@@ -134,15 +134,15 @@ func TestHashReaderVerification(t *testing.T) {
 		{
 			desc:       "Correct sha256, nested, truncated",
 			src:        mustReader(t, bytes.NewReader([]byte("abcd-more-stuff-to-be ignored")), 4, "", "", 4, false),
-			size:       0,
-			actualSize: 0,
+			size:       4,
+			actualSize: -1,
 			sha256hex:  "88d4266fd4e6338d13b845fcf289579d209c897823b9217da3e161936f031589",
 		},
 		{
 			desc:       "Correct sha256, nested, truncated, swapped",
-			src:        mustReader(t, bytes.NewReader([]byte("abcd-more-stuff-to-be ignored")), 0, "", "", 0, false),
+			src:        mustReader(t, bytes.NewReader([]byte("abcd-more-stuff-to-be ignored")), 4, "", "", -1, false),
 			size:       4,
-			actualSize: 4,
+			actualSize: -1,
 			sha256hex:  "88d4266fd4e6338d13b845fcf289579d209c897823b9217da3e161936f031589",
 		},
 		{
@@ -287,6 +287,14 @@ func TestHashReaderInvalidArguments(t *testing.T) {
 			size:       4,
 			actualSize: 4,
 			success:    true,
+		},
+		{
+			desc:        "Nested, size mismatch",
+			src:         mustReader(t, bytes.NewReader([]byte("abcd-morestuff")), 4, "", "", -1, false),
+			size:        2,
+			actualSize:  -1,
+			success:     false,
+			expectedErr: ErrSizeMismatch{Want: 4, Got: 2},
 		},
 	}
 

